### PR TITLE
refactor(cli): traversal

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

Part of #3307 

The PR breaks down the traversal into two parts:
- evaluation
- handling

The logic of the evaluation is pretty much the same we have today, and at the end instead of calling `handle_path`, we can a **new** function called `store_path`.

Once the evaluation finishes, we call `traversal` again, but we can a new function called `handle`, which essentially calls the existing function `process_file`.

Now that we have this new information, I took the opportunity to add two new **verbose** diagnostics that show the evaluated paths, and the paths that were changed. 

## Test Plan

Updated the current snapshosts.

### CODE REVIEW

File: crates/biome_cli/src/execute/traverse.rs [hunk 1]
```diff
@@ -483,11 +505,16 @@ pub(crate) struct TraversalOptions<'ctx, 'app> {
     /// The approximate number of diagnostics the console will print before
     /// folding the rest into the "skipped diagnostics" counter
     pub(crate) remaining_diagnostics: &'ctx AtomicU32,
+
+    /// List of paths that should be processed
+    pub(crate) evaluated_paths: RwLock<FxHashSet<EvaluatedPath>>,
```
Comment (Conaclos) [inline crates/biome_cli/src/execute/traverse.rs:510]: Just to be sure: why do we need a `RWLock` here? Have we several processes that “evaluate” files? Or it is just for toggling if a file was fixed? In this last case, we could use an atomic bool?

File: crates/biome_cli/src/execute/traverse.rs [hunk 2]
```diff
@@ -588,18 +619,25 @@ impl<'ctx, 'app> TraversalContext for TraversalOptions<'ctx, 'app> {
...
+    }
+
+    fn store_path(&self, path: &Path) {
+        self.evaluated_paths.write().unwrap().insert(path.into());
     }
 }
 
 /// This function wraps the [process_file] function implementing the traversal
 /// in a [catch_unwind] block and emit diagnostics in case of error (either the
 /// traversal function returns Err or panics)
```
Comment (Conaclos) [inline crates/biome_cli/src/execute/traverse.rs:633]: The docs should be updated

File: crates/biome_cli/src/reporter/mod.rs [hunk 3]
```diff
@@ -47,6 +49,16 @@ pub trait ReporterVisitor {
         summary: TraversalSummary,
     ) -> io::Result<()>;
 
+    /// Writes the paths that were handled during a run.
+    /// The fist list represents the paths that were evaluated, the second list represents the paths that were fixed
```
Comment (Conaclos) [inline crates/biome_cli/src/reporter/mod.rs:53]: Where is the second list here?

File: crates/biome_cli/tests/snapshots/main_cases_diagnostics/max_diagnostics_verbose.snap [hunk 4]
```diff
@@ -127,9 +115,41 @@ src/folder_0/package-lock.json project  VERBOSE  ━━━━━━━━━━
...
+  
+  - src/file.js
+  
+
+```
+
+```block
+ VERBOSE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+
+  i List of files that were fixed:
```
Comment (Conaclos) [inline crates/biome_cli/tests/snapshots/main_cases_diagnostics/max_diagnostics_verbose.snap:150]: We could simplify the wording to `List of fixed files:` or even `Fixed files:`/`Fixed:`

File: crates/biome_cli/tests/snapshots/main_cases_diagnostics/max_diagnostics_verbose.snap [hunk 5]
```diff
@@ -127,9 +115,41 @@ src/folder_0/package-lock.json project  VERBOSE  ━━━━━━━━━━
...
 
 ```block
 Checked 1 file in <TIME>. No fixes applied.
 Found 1 error.
 ```
+
+```block
+ VERBOSE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+
+  i List of files that were evaluated:
```
Comment (Conaclos) [inline crates/biome_cli/tests/snapshots/main_cases_diagnostics/max_diagnostics_verbose.snap:140]: We could simplify the wording to `List of evaluated files:` or even `Evaluated files:`/`Evaluated:`

File: crates/biome_diagnostics/src/advice.rs [hunk 6]
```diff
@@ -99,6 +99,28 @@ impl<T: Display> Advices for LogAdvice<T> {
...
+        if self.list.is_empty() {
+            visitor.record_log(LogCategory::Warn, &"The list is empty.")
+        } else {
+            let pattern_list: Vec<_> = self
+                .list
+                .iter()
+                .map(|pattern| pattern as &dyn Display)
+                .collect();
+
+            visitor.record_list(&pattern_list)
```
Comment (Conaclos) [inline crates/biome_diagnostics/src/advice.rs:119]: In another PR we could certainly accept an `impl Iterator` in order to avoid the allocation of a `Vec`.

File: crates/biome_fs/src/fs.rs [hunk 7]
```diff
@@ -323,7 +332,59 @@ pub trait TraversalContext: Sync {
...
+
+    pub fn is_fixed(&self) -> bool {
+        self.1
+    }
+
+    pub fn as_path(&self) -> &Path {
+        self.0.as_path()
+    }
+
+    pub fn to_bath_buf(&self) -> PathBuf {
```
Comment (Conaclos) [inline crates/biome_fs/src/fs.rs:373]: typo.

```suggestion
    pub fn to_path_buf(&self) -> PathBuf {
```

File: crates/biome_fs/src/fs.rs [hunk 8]
```diff
@@ -323,7 +332,59 @@ pub trait TraversalContext: Sync {
...
+    /// This method will be called by the traversal for each file it finds
+    /// where [TraversalContext::store_path] returned true
+    fn store_path(&self, path: &Path);
+
+    /// Returns the paths that should be handled
+    fn evaluated_paths(&self) -> FxHashSet<EvaluatedPath>;
+}
+
+#[derive(Debug, Eq, Clone)]
+pub struct EvaluatedPath(PathBuf, bool);
```
Comment (Conaclos) [inline crates/biome_fs/src/fs.rs:346]: I could opt for named struct instead:

```suggestion
pub struct EvaluatedPath{
  path: PathBuf,
  is_fixed: bool,
}
```

**Review by Conaclos — APPROVED** [review]
Looks good to me. I left some remarks and suggestions.

File: crates/biome_cli/src/execute/traverse.rs (hunk 1 above)
Comment (ematipico) [reply crates/biome_cli/src/execute/traverse.rs:510]: `TraversalOptions` is sent as a **reference** across multiple threads. So we need `RwLock` for two reasons:
- enable interior mutability when having a `&self`
- lock the resource when multiple threads want to update the list
